### PR TITLE
feat(keys): support secp256r1

### PR DESCRIPTION
**Pull request by LanfordCai**
### Motivation

This PR is to support `secp256r1` curve for `keys` package

### Solution

Add `ImportPrivateKey`, `GenerateKeypair`, `Sign` and `Verify` functions for `secp256r1` by using golang ecdsa package

### CODE REVIEW

**Comment by patrick-ogrady** [comment]
Thanks for putting this up, @LanfordCai! You'll need to sign your commits before I can merge this 😢

**Comment by LanfordCai** [comment]
Hi, @patrick-ogrady , I have signed my commit :-)

**Comment by patrick-ogrady** [comment]
Waiting for a review internally from some of our cryptography folks, then will merge (if no concerns)!

File: keys/keys.go
Comment (itstehkman) [inline keys/keys.go:111]: Just to double check, is this 32 bytes? That'll be good so that we have uniformity with secp256k1, in which the keypair.PrivateKey is 32 bytes.

Comment (itstehkman) [reply keys/keys.go:111]: Just checked myself, it is 32 bytes. https://play.golang.org/p/n0rt31c3lTs, so we are good here.

File: keys/signer_secp256r1_test.go
Comment (itstehkman) [inline keys/signer_secp256r1_test.go:48]: Is ECDSA Recovery truly not supported for secp256r1? Interesting.

File: keys/keys.go
Comment (mikelodder7) [inline keys/keys.go:98]: Check the `x`,`y` are on the curve and are not zero.

File: keys/signer_secp256r1.go
Comment (mikelodder7) [inline keys/signer_secp256r1.go:84]: check signature is not identity otherwise it will always be valid.

File: keys/signer_secp256r1.go
Comment (mikelodder7) [inline keys/signer_secp256r1.go:115]: Check that the key is not the identity element.

File: keys/keys.go
Comment (itstehkman) [reply keys/keys.go:111]: How about big endian vs little endian in secp256k1 vs secp256r1?

Comment (LanfordCai) [reply keys/keys.go:115]: Keys and signatures of both curves are encoded in big-endian form

File: keys/signer_secp256r1_test.go
Comment (LanfordCai) [reply keys/signer_secp256r1_test.go:48]: This means `type.EcdsaRecovery` sigType for secp256r1 is unsupported for rosetta-go-sdk, not `secp256r1`. ECDSA recovery **IS** supported for `secp256r1`, but his PR would not support this for some reasons:

1. Golang's `crypto/ecdsa` package doesn't provide a `SignRecoverable`  method for `secp256r1` to create recoverable signature({r, s, v}, v is the recovery id). And I didn't find some other long-tested packages provide this. 

2. I know ONT, NEO, ZEEPIN use normal `secp256r1` signatures in their transactions, but I don't know the use case of `secp256r1` recoverable signature. It seems many people just recover the PublicKey from {r, s}, for v is not necessary strictly speaking.

File: keys/keys.go
Comment (LanfordCai) [reply keys/keys.go:102]: fixed

File: keys/signer_secp256r1.go
Comment (LanfordCai) [reply keys/signer_secp256r1.go:129]: fixed

File: keys/signer_secp256r1.go
Comment (LanfordCai) [reply keys/signer_secp256r1.go:89]: This is handled by `ecdsa.Sign` method, please check: https://github.com/golang/go/blob/5a25a3fd1d8248c039e34f5ae02cc9c8198fc6d6/src/crypto/ecdsa/ecdsa.go#L262

File: keys/keys.go
Comment (mikelodder7) [reply keys/keys.go:102]: Also check that the points x and y are not zero. If they are, IsOnCurve still returns true but we don't want a public key as the identity point since any signature verification will return true.

File: keys/signer_secp256r1.go
Comment (mikelodder7) [reply keys/signer_secp256r1.go:89]: No that's not it. The signature is composed of two points, R and S. You should check that neither of these is equal to zero aka the identity element on the curve. If the signature is the identity element, it will always be valid.

File: keys/signer_secp256r1.go
Comment (mikelodder7) [reply keys/signer_secp256r1.go:129]: This checks that the points are on the curve which is excellent but need to also check for the identity element which is when x and y are zero.

File: keys/keys.go
Comment (LanfordCai) [reply keys/keys.go:102]: > // A Curve represents a short-form Weierstrass curve with a=-3.
//
// Note that the point at infinity (0, 0) is not considered on the curve, and
// although it can be returned by Add, Double, ScalarMult, or ScalarBaseMult, it
// can't be marshaled or unmarshaled, and IsOnCurve will return false for it.

This is the comment of crypto/elliptic, IsOnCurve will return false.

Comment (LanfordCai) [reply keys/keys.go:102]: Please check this: https://github.com/golang/go/blob/3298300ddf45a0792b4d8ea5e05f0fbceec4c9f9/src/crypto/elliptic/elliptic.go#L24 @mikebrowncb

Comment (mikebrowncb) [reply keys/keys.go:102]: 👋 

I think you meant to tag mikelodder

Comment (LanfordCai) [reply keys/keys.go:102]: Oh, I'm sorry

File: keys/signer_secp256r1.go
Comment (LanfordCai) [reply keys/signer_secp256r1.go:89]: The check for S is here: https://github.com/golang/go/blob/5a25a3fd1d8248c039e34f5ae02cc9c8198fc6d6/src/crypto/ecdsa/ecdsa.go#L272

File: keys/signer_secp256r1_test.go
Comment (itstehkman) [reply keys/signer_secp256r1_test.go:48]: Ok, good to know. Better to leave something out if not really used. Less code is more!

**Comment by patrick-ogrady** [comment]
@LanfordCai we are hoping to approve this soon! Sorry for the delay here.

File: keys/keys.go
Comment (itstehkman) [reply keys/keys.go:102]: Seems valid, and it's tested https://github.com/golang/go/blob/3298300ddf45a0792b4d8ea5e05f0fbceec4c9f9/src/crypto/elliptic/elliptic_test.go#L453

File: keys/signer_secp256r1.go
Comment (itstehkman) [inline keys/signer_secp256r1.go:35]: small nit: should this be shared with secp256k1? The signing algorithm ECDSA is constant between the two I believe, so these constants should be shared ideally. We can tackle in a future PR though, it's a small nit

File: keys/signer_secp256r1.go
Comment (itstehkman) [inline keys/signer_secp256r1.go:40]: What are you doing with this line? Seems strange and unused.

File: keys/signer_secp256r1.go
Comment (itstehkman) [reply keys/signer_secp256r1.go:129]: Per above, IsOnCurve checks this

File: keys/signer_secp256r1.go
Comment (patrick-ogrady) [reply keys/signer_secp256r1.go:40]: This is a common pattern in golang to [verify interface compliance](https://github.com/uber-go/guide/blob/master/style.md#verify-interface-compliance). We use this extensively throughout the SDK and CLI.

File: keys/signer_secp256r1.go
Comment (itstehkman) [inline keys/signer_secp256r1.go:47]: nit: document signature format. per below, it seems to be `R || S`

File: keys/signer_secp256r1.go
Comment (itstehkman) [reply keys/signer_secp256r1.go:89]: Yeah I think you're right, it seems to retry if R or S is 0, but will wait for Mike to ack here.

File: keys/signer_secp256r1.go
Comment (itstehkman) [reply keys/signer_secp256r1.go:40]: Makes sense. We should add a comment, it's confusing without context

File: keys/signer_secp256r1.go
Comment (LanfordCai) [reply keys/signer_secp256r1.go:35]: Yes, this is the same in secp256k1, so I named it as EcdsaRLen, not Secp256r1RLen. Maybe I should put this to https://github.com/coinbase/rosetta-sdk-go/blob/23200e54f95116600ff4617be5b78e1768ba689c/keys/signer_secp256k1.go#L33 ? But I don't want to modify secp256k1 code for this PR is for secp256r1, and this is not used in secp256k1 code.

File: keys/signer_secp256r1.go
Comment (LanfordCai) [reply keys/signer_secp256r1.go:41]: Added a comment for secp256r1 code. For secp256k1 and edwards25519 code, maybe it can be added in a futrue PR.

File: keys/signer_secp256r1.go
Comment (LanfordCai) [reply keys/signer_secp256r1.go]: Updated

File: keys/signer_secp256r1.go
Comment (itstehkman) [reply keys/signer_secp256r1.go:35]: That's fine, if it's not used in secp256k1 code then we can ignore it or tackle it later. It certainly is a great validation to check sig / message length

File: keys/signer_secp256r1.go
Comment (itstehkman) [reply keys/signer_secp256r1.go:91]: @mikelodder7 can you re-review?

File: keys/keys.go
Comment (itstehkman) [reply keys/keys.go:102]: @LanfordCai One more nit, let's just add a comment before the call to IsOnCurve to explicitly show that this is being checked since it is pretty important to know. Then I will +1

Comment (LanfordCai) [reply keys/keys.go:106]: @itstehkman I have added comments before each call to IsOnCurve

**Review by itstehkman — DISMISSED** [review]
Thanks @LanfordCai! This looks great 👏

**Comment by LanfordCai** [comment]
It requires all commits are signed, should I rebase and resign commits and force push to this branch?  @patrick-ogrady

**Comment by itstehkman** [comment]
@LanfordCai yes you should do that. I hope it doesn't wipe out the review history, I don't think it will. But let's see.

**Comment by LanfordCai** [comment]
@itstehkman I have resigned commits and the review history is still there

**Review by itstehkman — APPROVED** [review]

**Comment by patrick-ogrady** [comment]
🚀

**Pull request merged**
